### ISSUE TITLE
OutputAgent Digestion / Delay

### ISSUE THREAD

**Issue by level20peon** [cantino/huginn#1021]
As a former Yahoo Pipes user, I arranged my Agents' workflow similar to the Pipes Modules for most Scenarios I am using:

WebsiteAgent / RSSAgent (grab data) -> EventFormattingAgent (split/format data) -> TriggerAgent (remove certain items) -> DeDuplicationAgent (unify output) -> DataOutputAgent (output data as RSS feed)

Now there are data sources (Websites / RSS feeds) which constantly output new data, even in the range of minutes, so I had to set the WebSiteAgent / RSSAgent at the beginning of the workflow to run every few minutes (or so) in order to not miss any input.
However, I would prefer to have this data output once every hour or so, instead of having one new item every two minutes (or so).
Alternatively / additionally it would be great to have the OutputAgent "release" data if there are more than X items queued up.

Is it possible to implement such functionality with existing agents or would that be a new feature? If it would be a new feature, would it be feasible to implement? What do you guys thing?

**Comment by K1773R** [comment]
1) DelayAgent? wouldnt solve the problem...
2) RollUpAgent? RollUp like riemann.io would be cool.

**Comment by cantino** [comment]
Few questions:

1) If you have a DeDuplicationAgent, why does it matter how often it sources the feed?
2) Does the PeakDetector help?  It may not be sufficient, though.

**Comment by level20peon** [comment]
I use Thunderbird for mails, as a calendar and as an RSS Reader (via TinyTinyRSS). I often need to reply to mails in a timely fashion but when I switch to TB because of a "new mail" notification, I see that some feeds have a few new entries. So I go ahead and read them, instead of reading them once per hour or so.
I know that this could be alleviated by self control, just was wondering if someone else would find this useful.
By the way, I would call it SpoolerAgent.

**Comment by cantino** [comment]
I don't think this is possible yet (although probably could be done with the JavaScriptAgent).  It'd be an interesting contribution.

**Comment by level20peon** [comment]
ok, so how does the email digest agent "cache" entries?

**Comment by cantino** [comment]
It pushes incoming events into a queue [here](https://github.com/cantino/huginn/blob/master/app/models/agents/email_digest_agent.rb#L33), then pulls them back out when `check` is called [here](https://github.com/cantino/huginn/blob/master/app/models/agents/email_digest_agent.rb#L40).

**Comment by level20peon** [comment]
Couldn't that part of code just be copied to create an agent which queues events and propagates them to other agents based on the "Schedule" setting?

**Comment by cantino** [comment]
Yes, this is definitely possible if someone wants to work on it.  If you'd like to give it a try, I'm happy to help you get your environment setup.  You'll want to use git on the command line to make branches on your fork on Huginn so that you can make pull requests with multiple files.

**Comment by level20peon** [comment]
Which part of the code exactly propagates event output to the next agent? Could you point out an example in one of the existing agents?

**Comment by level20peon** [comment]
What do you think about this?

    module Agents
      class SpoolAgent < Agent
    
        default_schedule "every_12h"
    
        description <<-MD
          The Spool Agent collects any Events sent to it and outputs them all when scheduled.
        MD
    
        def receive(incoming_events)
          incoming_events.each do |event|
            self.memory['queue'] ||= []
            self.memory['queue'] << event.payload
            self.memory['events'] ||= []
            self.memory['events'] << event.id
          end
        end
    
        def check
          if self.memory['queue'] && self.memory['queue'].length > 0
            self.memory['events'].each do |event|
            #log("Propagating event '#{event}'.", :inbound_event => event )
            create_event payload: event
            self.memory['queue'] = []
            self.memory['events'] = []
          end
        end
      end
    end

**Comment by cantino** [comment]
You probably don't need both the `'queue'` and `'events'` sets.  I'd suggest keeping just `'events'` and using those ids in `check` to re-find the events and emit them.  Something like `events.where(id: memory['events'])`

But does it functionally work for you?

**Comment by level20peon** [comment]
I wanted to run it by you before I put some code in a language I don't speak onto my server :)

That being said, here's my next try:

    module Agents
      class SpoolerAgent < Agent
    
        default_schedule "every_12h"
    
        description <<-MD
          The Spooler Agent collects any Events sent to it and outputs them all when scheduled.
        MD
    
        def receive(incoming_events)
          incoming_events.each do |event|
            self.memory['events'] ||= []
            self.memory['events'] << event.id
          end
        end
    
        def check
          if self.memory['events'] && self.memory['events'].length > 0
            create_event payload: events.where(id: memory['events'])
            self.memory['events'] = []
          end
        end
    
      end
    end

btw: if you don't need to store the actual payloads in the memory, but can just refer to them by storing their ids, then why does the email digest agent actually store the payload in the memory, too?

**Comment by cantino** [comment]
You should definitely set Huginn up locally so that you can try this out in development.  Our README should take you through that process.

Do you want it to re-emit a single event, or all of the received ones?  If you wanted all of them, you could do:

```ruby
def check
  if self.memory['events'] && self.memory['events'].length > 0
    events.where(id: memory['events']).each do |event|
      create_event payload: event.payload
    end
    memory['events'] = []
  end
end
```

**Comment by level20peon** [comment]
I know how to do that and I meant that I don't want to throw anything in a language I don't speak, even onto a development server.
However, the agent doesn't work as it doesn't emit any events to agents down the line. Are you sure your idea about just storing the id is correct? If we just store the events' id then how can you access the payload of that id by just referring to the id, instead of the actual event somewhere inside the database?
Maybe that is the reason why the EmailDigestAgent stores both the id and payload in memory!?

**Comment by cantino** [comment]
Can I see your whole code?

**Comment by level20peon** [comment]
module Agents
      class SpoolerAgent < Agent
    
        default_schedule "every_12h"
    
        description <<-MD
          The Spooler Agent collects any id sent to it and outputs them all when scheduled.
        MD
    
        def default_options
          { "no options" => "are needed" }
        end
    
        def working?
          true
        end
    
        def receive(incoming_id)
          incoming_id.each do |event|
            self.memory['events'] ||= []
            self.memory['events'] << event.id
          end
        end
    
        def check
          if self.memory['events'] && self.memory['events'].length > 0
            events.where(id: memory['events']).each do |event|
              create_event payload: event.payload
            end
            memory['events'] = []
          end
        end
    
      end
    end

**Comment by cantino** [comment]
I think I see the problem.  `events` here is the current Agent's `events`, which won't contain the incoming ones.  Try `received_events.where(id: memory['events']).each do |event|`

Also, be sure to clear the memory of your agent in it's summary page if you change the innards.

Also, `receive` is getting an array of Events, not ids.  Try this:

```ruby
module Agents
  class BufferAgent < Agent
    default_schedule "every_12h"

    description <<-MD
      The Buffer Agent stores received Events and emits copies of them on a schedule.

      `expected_receive_period_in_days` is used to determine if the Agent is working. Set it to the maximum number of days
      that you anticipate passing without this Agent receiving an incoming Event.
    MD

    def default_options
      {
        'expected_receive_period_in_days' => "10"
      }
    end

    def validate_options
      unless options['expected_receive_period_in_days'].present? && options['expected_receive_period_in_days'].to_i > 0
        errors.add(:base, "Please provide 'expected_receive_period_in_days' to indicate how many days can pass before this Agent is considered to be not working")
      end
    end

    def working?
      last_receive_at && last_receive_at > options['expected_receive_period_in_days'].to_i.days.ago && !recent_error_logs?
    end

    def receive(incoming_events)
      incoming_events.each do |event|
        memory['events'] ||= []
        memory['events'] << event.id
      end
    end

    def check
      if memory['events'] && memory['events'].length > 0
        received_events.where(id: memory['events']).each do |event|
          create_event payload: event.payload
        end
        memory['events'] = []
      end
    end
  end
end
```

If this works, I can make a PR for it.  I tested and it works for me.

**Comment by level20peon** [comment]
Thanks for the help. I will only be able to test this once https://github.com/cantino/huginn/issues/1036 is resolved though.

**Comment by cantino** [comment]
I will be adding this Agent in #1055.

**Comment by cantino** [comment]
I've merged #1055 into master now.
